doc/api/async_hooks.md

```diff
@@ -950,6 +950,48 @@ If this method is called outside of an asynchronous context initialized by
 calling `asyncLocalStorage.run` or `asyncLocalStorage.runAndReturn`, it will
 return `undefined`.
 
+### `asyncLocalStorage.enterWith(store)`
+<!-- YAML
+added: REPLACEME
+-->
+
+* `store` {any}
+
+Calling `asyncLocalStorage.enterWith(store)` will transition into the context
+for the remainder of the current synchronous execution and will persist
+through any following asynchronous calls.
+
+Example:
+
+```js
+const store = { id: 1 };
+asyncLocalStorage.enterWith(store);
+asyncLocalStorage.getStore(); // Returns the store object
+someAsyncOperation(() => {
+  asyncLocalStorage.getStore(); // Returns the same object
+});
+```
+
+This transition will continue for the _entire_ synchronous execution.
+This means that if, for example, the context is entered within an event
+handler subsequent event handlers will also run within that context unless
+specifically bound to another context with an `AsyncResource`.
+
+```js
+const store = { id: 1 };
+
+emitter.on('my-event', () => {
+  asyncLocalStorage.enterWith(store);
+});
+emitter.on('my-event', () => {
+  asyncLocalStorage.getStore(); // Returns the same object
+});
+
+asyncLocalStorage.getStore(); // Returns undefined
+emitter.emit('my-event');
+asyncLocalStorage.getStore(); // Returns the same object
```

Comment:
So, with this example, I am not sure it is very safe to introduce `enterWith()` without an simmetric method to exit the context.  One should alwys know and control which chunks of code have access to the store or not.

But we know we have several issue with such flow. At this point, we might want to put this PR on hold for now and iterate when we have feedback. wdyt?

Reply:
> One should alwys know and control which chunks of code have access to the store or not.

To my understanding the use case for this that some APM tool sets it's context via `enterSync` in the first listener of some new incoming request. This context is then also active in the application code (which is not aware at all of this AsyncLocalStorage instance) without the need to wrap the other listeners via `run`. So the context is not leaking per accident here, it's by intention and one of the main reasons to add it.

I agree that this use case is not shown exactly by the sample. But it's not that easy to describe this to non APM aware people in a few markdown lines.

Reply:
A triggering event (for instance, http request, lambda invokation or db call) can trigger several codepaths at the same time. It might make sense not to link one of them to the event on purpose. Wether it is as a defensive feature (I don't want to follow a codepath that might not fill the finish condition I am expecting) or because it might make little sense (Let's say a user uses the trigger event to start some code that is unrelated to the event itself, like a hacked around crontab), in that case, we would want to avoid associating the context to some codepath. Doing otherwise could lead to an increase of the memory pressure on the app by potentially delaying the Garbage Collection of a context.

Reply:
I agree that unwanted propagation may happen. But this happens also if the same tool wraps the application HTTP request listener in a `AsyncLocalStorage.run()` call which is what APM tools are doing now. In general APM tools want to monitor "everything" triggered within the root request.

We have no mechanism in place to stop propagation based on e.g. the type of an async resource created (usually you don't want to propagate into an interval time as may runs forever).

Reply:
In APM, the "finish condition" of the trigger event is not always _known_. That's kind of _the point_ of this PR. It is to say "I want everything following this point in the async call graph to be attributed to my context and I'll sort it out later"

I'm _totally_ happy with lots of docs warnings about the implications for regular users wanting to use this API, but APM _needs_ this. The `run(...)` method is _not_ sufficient to eliminate patching, which is the whole _point_ of async_hooks, and everything built on it, in the first place.

Reply:
Also, for _escaping_ the context, we can just use the `exit(...)` methods as-needed.

Reply:
> Also, for escaping the context, we can just use the exit(...) methods as-needed.

Not really, the `exit*(...)` exit a given functionnal scope. Let's say in a single synchronous block, I want a few lines to be in context and some to be out. How can I do that using `enterWith(...)`.

> The run(...) method is not sufficient to eliminate patching, which is the whole point of async_hooks, and everything built on it, in the first place.

Is it possible to use `enterWith(...)` in an HTTP server as an APM/RASP without patching something in `http.Server` ?

Reply:
You just use `exit*(...)` like you normally would? :man_shrugging: 

```js
// not in context
cls.enterWith({})
// in context
cls.exit(() => {
  // not in context
})
// in context
```

As for patching `http.Server`, it depends a bit what data you want to collect, but yes, it's possible to do quite a bit with zero patching, so long as you can inject the context like this. For example:

```js
server.on('request', (request, response) => {
  const transaction = new Transaction(req.url)
  cls.enterWith(transaction)
  response.on('finish', () => {
    transaction.report()
  })
})
```

Reply:
It still feel weird to me that I could not do:
```
// not in context
cls.enterWith({})
// in context
cls.exitSync()
// not in context
```
wdyt?

> As for patching http.Server, it depends a bit what data you want to collect, but yes, it's possible to do quite a bit with zero patching, so long as you can inject the context like this.
Good point!

Reply:
The `exitSync()` idea is specifically where most of the problems with domains came from, so I'm rather hesitant to approve that. However, `AsyncLocalStorage` _is_ more self-contained and _hopefully_ that could _extend_ to a rewrite of domains sufficiently that the enter/exit controls would not be exposed to someone that shouldn't have that access. I'm a bit conflicted about it at the moment. :thinking: 

What do you think of going forward with `enterWith(...)` for now and evaluating separately if we want to do something about a sync exit?

Reply:
@Qard sure! ATM, I don't see any strong blocker on this PR (pretty much why I don't have a blocking review on it ^^).

Reply:
@Qard @vdeturckheim I'd like to propose at least to rename `enterWith` into something like `setStore` (as far as I remember @vdeturckheim already came up with this name in one of the previous PRs). Current name will confuse users, as `AsyncLocalStorage` will be having `enterWith` and `exit` which sound like a symmetrical pair of methods, while they're not. Also, `setStore` will make a natural pair with `getStore`. WDYT?

Reply:
On the other hand `enterWith` does more than a setter method, so I'm not that sure. Still, I think, it could be given a better name that would help with avoid user confusion.

Reply:
I'd be -1 on `setStore`, as you mentionned, it is not a setter method but really a method that change the state of the object. The nature of a held payload must not impact the behavior of the holding mechanism.

Reply:
@vdeturckheim I still think that there should be a better name for this method. Logical names are important for an API.

Reply:
I think it's like a `set` if you are within a `run` but an `enter` if not. Hard to come up with a name here as the interactions between nested calls are hard to understand.
I agree that the postfix `with` is not needed.